sw/device/lib/dif/dif_usbdev.c

```diff
@@ -0,0 +1,1061 @@
+  if (usbdev == NULL || dst == NULL) {
+    return kDifUsbdevRxPacketReadBytesResultBadArg;
+  }
+
+  // Make sure that a read buffer operation is in progress
+  if (!buffer_op_is_reading(usbdev)) {
+    return kDifUsbdevRxPacketReadBytesResultNotReading;
+  }
+  // bytes_to_copy is the minimum of remaining bytes and `dst_len`
+  size_t bytes_to_copy = usbdev->read_buffer_op_state.remaining_bytes;
+  if (bytes_to_copy > dst_len) {
+    bytes_to_copy = dst_len;
+  }
+  // Copy payload to buffer
+  const uint32_t buffer_addr =
+      get_buffer_addr(usbdev->read_buffer_op_state.buffer_entry,
+                      usbdev->read_buffer_op_state.offset);
+  mmio_region_memcpy_from_mmio32(usbdev->base_addr, buffer_addr, dst,
+                                 bytes_to_copy);
+  // Update current buffer operation
+  usbdev->read_buffer_op_state.offset += bytes_to_copy;
+  usbdev->read_buffer_op_state.remaining_bytes -= bytes_to_copy;
+  // Check if the entire payload is read
+  dif_usbdev_rx_packet_read_bytes_result_t res =
+      kDifUsbdevRxPacketReadBytesResultContinue;
+  if (usbdev->read_buffer_op_state.remaining_bytes == 0) {
+    res = kDifUsbdevRxPacketReadBytesResultOK;
+    // Finish read buffer operation
+    buffer_op_stop_read(usbdev);
+    // Return the buffer to the free buffer pool
+    buffer_pool_add(&usbdev->buffer_pool,
+                    usbdev->read_buffer_op_state.buffer_entry);
+  }
+
+  if (bytes_written != NULL) {
+    *bytes_written = bytes_to_copy;
+  }
+
+  return res;
+}
+
+dif_usbdev_rx_packet_discard_bytes_result_t dif_usbdev_rx_packet_discard_bytes(
+    dif_usbdev_t *usbdev) {
+  if (usbdev == NULL) {
+    return kDifUsbdevRxPacketDiscardBytesResultBadArg;
+  }
+
+  // Make sure that a read buffer operation is in progress
+  if (!buffer_op_is_reading(usbdev)) {
+    return kDifUsbdevRxPacketDiscardBytesResultNotReading;
+  }
+  // Finish read buffer operation
+  buffer_op_stop_read(usbdev);
+  // Return the buffer to the free buffer pool
+  buffer_pool_add(&usbdev->buffer_pool,
+                  usbdev->read_buffer_op_state.buffer_entry);
+
+  return kDifUsbdevRxPacketDiscardBytesResultOK;
+}
+
+dif_usbdev_tx_packet_write_bytes_result_t dif_usbdev_tx_packet_write_bytes(
+    dif_usbdev_t *usbdev, uint8_t endpoint, uint8_t *src, size_t src_len,
+    size_t *bytes_written) {
+  if (usbdev == NULL || (src_len > 0 && src == NULL) ||
+      !is_valid_endpoint(endpoint)) {
+    return kDifUsbdevTxPacketWriteBytesResultBadArg;
+  }
+
+  // Get tx packet status for the endpoint
+  dif_usbdev_tx_packet_status_t in_packet_status;
+  if (dif_usbdev_tx_packet_get_status(usbdev, endpoint, &in_packet_status) !=
+      kDifUsbdevOK) {
+    return kDifUsbdevTxPacketWriteBytesResultError;
+  }
+
+  // HW details
+  const uint32_t config_in_reg_offset =
+      kEndpointHwInfos[endpoint].config_in_reg_offset;
+  uint32_t config_in_val;
+  uint8_t buffer_entry;
+  uint8_t buffer_len;
+
```

Comment:
Nit: bring the three predeclared variables down one line, so they're part of the `switch` stanza that initializes them.